tests/subsys/canbus/isotp/implementation/src/main.c

```diff
@@ -0,0 +1,446 @@
+	struct net_buf *buf, *frag;
+	const u8_t *data_ptr;
+
+	ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+			  &fc_opts_single, K_NO_WAIT);
+	zassert_equal(ret, 0, "Binding failed (%d)", ret);
+
+	for (i = 0; i < NUMBER_OF_REPETITIONS; i++) {
+		send_test_data(can_dev, random_data, send_len);
+		data_ptr = random_data;
+
+		ret = iso_tp_recv_net(&recv_ctx, &buf, K_MSEC(1000));
+		zassert_equal(ret, 0, "recv returned %d", ret);
+		buf_len = net_buf_frags_len(buf);
+		zassert_equal(buf_len, send_len, "Data length differ");
+		frag = buf;
+
+		do {
+			data_ptr = check_frag(frag, data_ptr);
+			memset(frag->data, 0, frag->len);
+		} while ((frag = frag->frags));
+
+		net_buf_unref(buf);
+	}
+
+	iso_tp_unbind(&recv_ctx);
+}
+
+static void test_send_receive_single_block(void)
+{
+	const size_t send_len = CONFIG_CAN_ISOTP_RX_BUF_COUNT *
+				CONFIG_CAN_ISOTP_RX_BUF_SIZE + 6;
+	int ret, i;
+
+	ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+			  &fc_opts_single, K_NO_WAIT);
+	zassert_equal(ret, 0, "Binding failed (%d)", ret);
+
+	for (i = 0; i < NUMBER_OF_REPETITIONS; i++) {
+		send_test_data(can_dev, random_data, send_len);
+
+		memset(data_buf, 0, sizeof(data_buf));
+		ret = iso_tp_recv(&recv_ctx, data_buf, sizeof(data_buf),
+				  K_MSEC(1000));
+		zassert_equal(ret, send_len,
+			      "data should be received at once (ret: %d)", ret);
+		ret = memcmp(random_data, data_buf, send_len);
+		zassert_equal(ret, 0, "Data differ");
+	}
+
+	iso_tp_unbind(&recv_ctx);
+}
+
+static void test_bind_unbind(void)
+{
+	int ret, i;
+
+	for (i = 0; i < 100; i++) {
+		ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+				  &fc_opts, K_NO_WAIT);
+		zassert_equal(ret, 0, "Binding failed (%d)", ret);
+		iso_tp_unbind(&recv_ctx);
+	}
+
+	for (i = 0; i < NUMBER_OF_REPETITIONS; i++) {
+		ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+				  &fc_opts, K_NO_WAIT);
+		zassert_equal(ret, 0, "Binding failed (%d)", ret);
+		send_sf(can_dev);
+		get_sf_net(&recv_ctx);
+		iso_tp_unbind(&recv_ctx);
+	}
+
+	for (i = 0; i < NUMBER_OF_REPETITIONS; i++) {
+		ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+				  &fc_opts, K_NO_WAIT);
+		zassert_equal(ret, 0, "Binding failed (%d)", ret);
+		send_sf(can_dev);
+		get_sf(&recv_ctx);
+		iso_tp_unbind(&recv_ctx);
+	}
+
+	for (i = 0; i < 10; i++) {
+		ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+				  &fc_opts, K_NO_WAIT);
+		zassert_equal(ret, 0, "Binding failed (%d)", ret);
+		send_test_data(can_dev, random_data, 60);
+		receive_test_data_net(&recv_ctx, random_data, 60, 0);
+		iso_tp_unbind(&recv_ctx);
+	}
+
+	for (i = 0; i < NUMBER_OF_REPETITIONS; i++) {
+		ret = iso_tp_bind(&recv_ctx, can_dev, &rx_addr, &tx_addr,
+				  &fc_opts, K_NO_WAIT);
+		zassert_equal(ret, 0, "Binding failed (%d)", ret);
+		send_test_data(can_dev, random_data, 60);
+		receive_test_data(&recv_ctx, random_data, 60, 0);
+		iso_tp_unbind(&recv_ctx);
+	}
+}
+
+static void test_multi_instances(void)
+{
+
```

Comment:
I guess this test could be removed if empty?

Reply:
Removed.